### DISCUSSION TITLE
Use libssl's base64 encoding/decoding instead of unmaintained libb64 ?

### DISCUSSION THREAD

**Discussion by kapouer** [transmission/transmission#6970] [Ideas]
Hi,

it seems EVP_ interfaces in openssl can deal with base64 encoding/decoding.

**Comment by killemov** [comment]
If we are already using libssl with Transmission anyway then yes else maybe. Is libb64 really unmaintained or just done?

**Reply by kapouer** [reply]
AFAICT it is forked and embedded in Transmission. It doesn't match the upstream version anymore.

**Comment by tearfur** [comment]
I'm not too convinced about that idea.

I found a benchmark done by the R project suggesting that openssl's base64 implementation is slow: https://cran.r-project.org/web/packages/base64url/vignettes/Benchmarks.html. I'm actually surprised by it, I thought if anyone would have a well-tested blazing-fast assembly implementation, it would be them.

On the other hand, here is a benchmarking project comparing different C/C++ libraries https://github.com/gaspardpetit/base64. If we are going to change the base64 implementation, it looks like https://github.com/aklomp/base64 is the fastest one by far.

**Reply by kapouer** [reply]
Is base64 heavily used in bittorrent protocol ?

**Reply by tearfur** [reply]
No, the BT protocol itself doesn't use it (in our implementation at least).

**Reply by kapouer** [reply]
So it's just used for magnet links or what ? The trade-off between security and maintainability or performance is not worth it, is it ?

**Reply by tearfur** [reply]
> So it's just used for magnet links or what ?

Base64 is currently only used for logging binary strings, or transporting them via the RPC interface, and writing tests.

> The trade-off between security and maintainability or performance is not worth it, is it ?

It's just that I prefer to keep it as-is if it works without problems.

What is this "security" that we are trading off by choosing to keep libb64 anyway? Base64 is not rocket science, it is literally just a one-to-one mapping of 6-bit bitfields to ASCII characters, and vice versa. So it is very possible for an implementation to just be "done".

**Comment by tearfur** [comment]
Turns out we *were* using the base64 functions from crypto libraries, but @mikedld decided to switch to libb64 at b928ae1ba824426385e6a078f7d84f1452e9b370.
